### ISSUE TITLE
Performance declines when using tune_grid for nested cross validation

### ISSUE THREAD

**Issue by ercbk** [tidymodels/tune#148]
## The problem

Hi, I'm trying out different implementations of nested cross-validation. I've coded one that was adapted from the Nested Resampling [article](https://tidymodels.github.io/rsample/articles/Applications/Nested_Resampling.html) in the {rsample} package and one that uses `tune_grid` from the {tune} package. I've noticed significant differences in the performances of each. 

After reading the Optimizations and Parallel Processing [article](https://tidymodels.github.io/tune/articles/extras/optimizations.html#fn1), I thought it might have something to do with whether the parallelization happens in the folds or the resamples, but I'm pretty sure both parallelize in the resamples. From what I can tell, the difference in performance is related to using parsnip to encode the models and something that's happening in tune_grid. When running the tune_grid method for the rf, this message repeats, "i Creating pre-processing data to finalize unknown parameter: mtry." So maybe there's something extra happening here.

For 100 observations and rf and svm algorithms with 202x2 and 11x1 hyperparameter grids respectively, the complete nested-cv process for the tune_grid implementation (method 1) took 5.27 hrs while using the other way (method 2) took 3.66 hrs. I used 5 repeats of 10 folds for the outer loop and 25 resamples in each fold for the parameter-tuning, inner-loop.  I'm using a 4 core, 16G RAM WIndows 10 machine.

The code I've included only runs 2 folds of the inner-loop. The rest of the code doesn't seem to be very expensive, but if you want, I can include it this thread or in some gists.
Here are some of times:

* Using tune_grid (method 1)
      + parsnip:ranger:::rf = 657.17 sec elapsed (11 min)
      + parsnip::kernlab::svm = 37.36 sec elapsed sec elapsed
 
* Method 2
      + parsnip::ranger::rf = 571.75 sec elapsed (9.5 min)
      + kernlab::svm = 3.46 sec elapsed
      + parsnip::kernlab::svm = 33.44 sec elapsed
 
(aside: Method 2 times had some variance and ranged from around 462 sec to 656 sec depending on if I ran the code directly after method 1 or not.)

I wondering if this is something in my code or some option I'm not including or if this just a gradual accumulation from grid searching 1250 models. I'd appreciate any help you might offer. Thank you.

## Reproducible example

``` r
#####################################
# Set-up
#####################################


pacman::p_load(doFuture, dplyr, purrr, tidymodels, tune)

set.seed(2019)

registerDoFuture()
plan(multiprocess)

# simulated data; generates 10 multi-patterned, numeric predictors plus outcome variable
sim_data <- function(n) {
      tmp <- mlbench::mlbench.friedman1(n, sd=1)
      tmp <- cbind(tmp$x, tmp$y)
      tmp <- as.data.frame(tmp)
      names(tmp)[ncol(tmp)] <- "y"
      tmp
}

# Use small data to tune and compare models
small_dat <- sim_data(100)

ncv_dat <- nested_cv(small_dat, 
                     outside = vfold_cv(repeats = 5), 
                     inside = bootstraps(times = 25)) %>% 
      slice(1:2)




######################################
# method 1: Using Tune pkg 
######################################


error_funs <- metric_set(mae)

rf_params <- crossing("mtry" = c(3,4), "trees" = seq(200, 300))

rf_inner <- rand_forest(mtry = tune(), trees = tune()) %>%
      set_engine("ranger", importance = 'impurity') %>%
      set_mode("regression")

svm_params <- tibble(cost = 2 ^ seq(-2, 8, by = 1))

svm_inner <- svm_rbf(cost = tune()) %>%
      set_engine("kernlab") %>%
      set_mode("regression")


tictoc::tic()
tuning_results <- map(ncv_dat$inner_resamples, function(dat, mod_inner, params) {
      tune_grid(y ~ .,
                model = mod_inner,
                resamples = dat,
                grid = params,
                metrics = error_funs)
},
rf_inner, rf_params)
#> i Creating pre-processing data to finalize unknown parameter: mtry
#> i Creating pre-processing data to finalize unknown parameter: mtry
tictoc::toc()
#> 667.56 sec elapsed


tictoc::tic()
tuning_results <- map(ncv_dat$inner_resamples, function(dat, mod_inner, params) {
      tune_grid(y ~ .,
                model = mod_inner,
                resamples = dat,
                grid = params,
                metrics = error_funs)
},
svm_inner, svm_params)
tictoc::toc()
#> 40.18 sec elapsed




############################################
# method 2: Adapted from "Nested Resampling"
############################################


error_FUN <- function(y_obs, y_hat){
      y_obs <- unlist(y_obs)
      y_hat <- unlist(y_hat)
      Metrics::mae(y_obs, y_hat)
}

rf_FUN <- function(params, analysis_set) {
      mtry <- params$mtry[[1]]
      trees <- params$trees[[1]]
      model <- parsnip::rand_forest(mode = "regression", mtry = mtry, trees = trees) %>%
            parsnip::set_engine("ranger", importance = 'impurity') %>%
            generics::fit(y ~ ., data = analysis_set)
      model
}

params_list <- list(rf = rf_params)
mod_FUN_list <- list(rf = rf_FUN)


# svm_FUN <- function(params, analysis_set) {
#       cost <- params$cost[[1]]
#       model <- kernlab::ksvm(y ~ ., data = analysis_set,  C = cost)
#       model
# }
# 
# svm_FUN <- function(params, analysis_set) {
#       cost <- params$cost[[1]]
#       parsnip::svm_rbf(cost = cost) %>%
#             parsnip::set_engine("kernlab") %>%
#             parsnip::set_mode("regression") %>%
#             generics::fit(y ~ ., data = analysis_set)
# }
# 
# params_list <- list(svm = svm_params)
# mod_FUN_list <- list(svm = svm_FUN)


# inputs params, model, and resample, calls model and error functions, outputs error
mod_error <- function(params, mod_FUN, dat) {
      y_col <- ncol(dat$data)
      y_obs <- assessment(dat)[y_col]
      mod <- mod_FUN(params, analysis(dat))
      pred <- predict(mod, assessment(dat))      
      error <- error_FUN(y_obs, pred)
      error
}

# inputs resample, loops hyperparam grid values to model/error function, collects error value for hyperparam combo
tune_over_params <- function(dat, mod_FUN, params) {
      params$error <- map_dbl(1:nrow(params), function(row) {
            params <- params[row,]
            mod_error(params, mod_FUN, dat)
      })
      params
}

# inputs and sends fold's resamples to tuning function, collects and averages fold's error for each hyperparameter combo
summarize_tune_results <- function(dat, mod_FUN, params) {
      # Return row-bound tibble that has the 25 bootstrap results
      param_names <- names(params)
      furrr::future_map_dfr(dat$splits, tune_over_params, mod_FUN, params) %>%
            # For each value of the tuning parameter, compute the
            # average <error> which is the inner bootstrap estimate.
            group_by_at(vars(param_names)) %>%
            summarize(mean_error = mean(error, na.rm = TRUE),
                      n = length(error))
}


tictoc::tic()
algorithm_comparison <- map2(mod_FUN_list, params_list, function(mod_FUN, params){
      # tune models by grid searching on resamples in the inner loop (e.g. 5 repeats 10 folds = list of 50 tibbles with param and mean_error cols)
      tuning_results <- map(ncv_dat$inner_resamples, summarize_tune_results, mod_FUN, params)
})
tictoc::toc()
#> 652.39 sec elapsed
```

<sup>Created on 2020-01-05 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by ercbk** [comment]
Session Info

R version 3.6.2 (2019-12-12)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18362)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] tune_0.0.1       yardstick_0.0.4  tibble_2.1.3     rsample_0.0.5    tidyr_1.0.0      recipes_0.1.8   
 [7] parsnip_0.0.4    infer_0.5.1      ggplot2_3.2.1    dials_0.0.4      scales_1.1.0     broom_0.5.3     
[13] tidymodels_0.0.3 purrr_0.3.3      dplyr_0.8.3      doFuture_0.8.2   iterators_1.0.12 foreach_1.4.7   
[19] future_1.15.1    globals_0.12.5  

loaded via a namespace (and not attached):
  [1] minqa_1.2.4         colorspace_1.4-1    class_7.3-15        ggridges_0.5.1      rsconnect_0.8.16   
  [6] markdown_1.1        base64enc_0.1-3     tidytext_0.2.2      rstudioapi_0.10     listenv_0.8.0      
 [11] furrr_0.1.0         rstan_2.19.2        SnowballC_0.6.0     DT_0.11             prodlim_2019.11.13 
 [16] fansi_0.4.0         lubridate_1.7.4     codetools_0.2-16    splines_3.6.2       knitr_1.26         
 [21] shinythemes_1.1.2   zeallot_0.1.0       bayesplot_1.7.1     nloptr_1.2.1        pROC_1.15.3        
 [26] workflows_0.1.0     shiny_1.4.0         compiler_3.6.2      tictoc_1.0          backports_1.1.5    
 [31] assertthat_0.2.1    Matrix_1.2-18       fastmap_1.0.1       lazyeval_0.2.2      cli_2.0.0          
 [36] later_1.0.0         htmltools_0.4.0     prettyunits_1.0.2   tools_3.6.2         igraph_1.2.4.2     
 [41] gtable_0.3.0        glue_1.3.1          reshape2_1.4.3      Rcpp_1.0.3          DiceDesign_1.8-1   
 [46] vctrs_0.2.1         nlme_3.1-143        crosstalk_1.0.0     timeDate_3043.102   xfun_0.11          
 [51] gower_0.2.1         stringr_1.4.0       ps_1.3.0            lme4_1.1-21         mime_0.8           
 [56] miniUI_0.1.1.1      lifecycle_0.1.0     pacman_0.5.1        gtools_3.8.1        tidypredict_0.4.3  
 [61] MASS_7.3-51.4       zoo_1.8-6           ipred_0.9-9         rstanarm_2.19.2     colourpicker_1.0   
 [66] promises_1.1.0      inline_0.3.15       shinystan_2.5.0     tidyposterior_0.0.2 gridExtra_2.3      
 [71] loo_2.2.0           StanHeaders_2.19.0  rpart_4.1-15        stringi_1.4.3       tokenizers_0.2.1   
 [76] dygraphs_1.1.1.6    lhs_1.0.1           boot_1.3-24         pkgbuild_1.0.6      lava_1.6.6         
 [81] rlang_0.4.2         pkgconfig_2.0.3     matrixStats_0.55.0  lattice_0.20-38     rstantools_2.0.0   
 [86] htmlwidgets_1.5.1   tidyselect_0.2.5    processx_3.4.1      plyr_1.8.5          magrittr_1.5       
 [91] R6_2.4.1            generics_0.0.2      pillar_1.4.3        withr_2.1.2         xts_0.11-2         
 [96] survival_3.1-8      nnet_7.3-12         janeaustenr_0.1.5   crayon_1.3.4        mlbench_2.1-1      
[101] grid_3.6.2          callr_3.4.0         threejs_0.3.1       digest_0.6.23       xtable_1.8-4       
[106] httpuv_1.5.2        stats4_3.6.2        GPfit_1.0-8         munsell_0.5.0       shinyjs_1.0

**Comment by topepo** [comment]
Sorry for the delay. It's a complicated issue. 

I've thought about this a bit and have reproduced your timings (at least the ratios). I don't have a better solution (right now) and don't see anything wrong with how you have approached it. 

Some general thoughts: 

 * It would make sense in method 1 to parallelize the _outer_ loop instead of the inner loop since that is the one that is active the longest. To do this you can use `furrr` (see below). For this example (with two outer resamples) it was the slowest but that might not be the case in general. It was 4-fold slower than method 1 and this is because it can only do two things at a time in parallel. 

```r
# method 1.5: 
library(furrr)
tictoc::tic()
tuning_results <- future_map(ncv_dat$inner_resamples, function(dat, mod_inner, params) {
  tune_grid(y ~ .,
            model = mod_inner,
            resamples = dat,
            grid = params,
            metrics = error_funs,
            control = control_grid(allow_par = FALSE))
},
rf_inner, rf_params)
tictoc::toc()
```

* In the spirit of WCGW, I ran method 1.5 without `allow_par = FALSE`. `foreach`/`future` parallelized only the outer loop and did _not_ start up 2 x 25 worker processes. 

 * We are looking to have an alternative parallelization scheme that _flattens_ the loops. Right now, you can think of what `tune_grid()` does as being three nested loops: resamples then preprocessing then modeling. Currently we only parallelize the resamples because we want to avoid re-creating the pre-processing when we don't have to. However, if the pre-processing is easy/fast, this is sub-optimal. The new scheme would parallelize a single (resample 1/recipe 1/model 1 then  resample 1/recipe 1/model 2 etc). Once we have this other scheme as an option, these timings might change for the better.

* None of the models that you listed use `multi_predict()` and `tune_grid()` exploits this. With other models, you might have different results. 

* Similarly, tuning a recipe would complicate this.

* method 2 being better might imply that the overhead/generality of `tune_grid()` is the source of inefficiency. I don't think that this would be the only source of inefficiency though. 

* I do have plans on writing a tune function for nested resampling. It is lower priority though.

**Comment by ercbk** [comment]
Thanks for the reply, Max. Yeah, I tried to parallelize the outer-loop, and I couldn't get it to work either. Doesn't make sense. I'm thinking about trying to make a nested-cv object for time series if I ever get (sort of) caught-up with some other projects. I haven't looked at the guts of your `nested_cv` function, but do you see any problems with using that as a framework?
